crates/sui-core/src/gateway_state.rs

```diff
@@ -368,12 +374,14 @@ where
 
     async fn get_sui_object(&self, object_id: &ObjectID) -> Result<SuiObject, anyhow::Error> {
         let object = self.get_object_internal(object_id).await?;
-        self.to_sui_object(object)
+        self.to_sui_object(object).await
     }
 
-    fn to_sui_object(&self, object: Object) -> Result<SuiObject, anyhow::Error> {
-        let cache = ModuleCache::new(&*self.store);
-        let layout = object.get_layout(ObjectFormatOptions::default(), &cache)?;
+    async fn to_sui_object(&self, object: Object) -> Result<SuiObject, anyhow::Error> {
+        let layout = object.get_layout(
+            ObjectFormatOptions::default(),
+            &*self.module_cache.lock().await,
```

Comment:
the reason of the mutex lock is to make the resolver sync+send, which is required by GatewayState impl... 
I have just found out there is a `SyncModuleCache` which impl sync, I will use that instead of using Mutex